File: problem1-50/problem18/problem18.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
using namespace std;
const int n = 15;
int f[n + 1][n + 1];
int main()
{
	freopen("problem18.txt", "r", stdin);
	for (int i = 1; i <= n; ++ i)
		for (int j = 1; j <= i; ++ j)
			cin >> f[i][j];
	for (int i = 2; i <= n; ++ i)
	{
		f[i][1] += f[i - 1][1];
		f[i][i] += f[i - 1][i - 1];
		for (int j = 2; j < i; ++ j)
			f[i][j] += max(f[i - 1][j - 1], f[i - 1][j]);
	}
	int ans = 0;
	for (int i = 1; i <= n; ++ i)
		ans = max(ans, f[n][i]);
	cout << ans << endl;
	return 0;
}

File: problem1-50/problem18/problem18.pas
program problem18;

const maxn=100;
var a,f:array[1..maxn,1..maxn] of longint;

    i,j,n:longint;

function max(a,b:longint):longint;

 begin

     if a>b then max:=a else max:=b;

 end;


begin

    assign(input,'problem18.txt'); reset(input);

    readln(n);

    for i:=1 to n do

       for j:=1 to i do read(a[i,j]);

    close(input);

    fillchar(f,sizeof(f),0);

    for i:=n downto 1 do

       for j:=1 to i do

         f[i,j]:=max(f[i+1,j],f[i+1,j+1])+a[i,j];

    writeln(f[1,1]);

end.